Dockerfile

```diff
@@ -27,11 +27,57 @@ RUN mkdir -p rootfs/"${CRYPTOLPATH}" \
 USER root
 RUN chown -R root:root /cryptol/rootfs
 
+
+from debian:stretch AS solvers
+
+# Install needed packages for building
+RUN apt-get update \
+    && apt-get install -y curl cmake gcc g++ git libreadline-dev
+RUN useradd -m user
+RUN install -d -o user -g user /solvers
+USER user
+WORKDIR /solvers
+RUN mkdir -p rootfs/usr/local/bin
+
+# Build abc from Galois' fork
+RUN git clone https://github.com/GaloisInc/abc.git
+RUN cd abc && make -j$(nproc)
+RUN cp abc/abc rootfs/usr/local/bin
+
+# Build Boolector release 3.2.1 from source
+RUN curl -L https://github.com/Boolector/boolector/archive/3.2.1.tar.gz | tar xz
+RUN cd boolector* && ./contrib/setup-lingeling.sh && ./contrib/setup-btor2tools.sh && ./configure.sh && cd build && make -j$(nproc)
+RUN cp boolector*/build/bin/boolector rootfs/usr/local/bin
+
+# Install yices 2.6.2
+RUN curl -L https://yices.csl.sri.com/releases/2.6.2/yices-2.6.2-x86_64-pc-linux-gnu-static-gmp.tar.gz | tar xz
+RUN cp yices*/bin/yices-smt2 rootfs/usr/local/bin
+
+# Install cvc4 1.8
+RUN curl -L https://github.com/CVC4/CVC4/releases/download/1.8/cvc4-1.8-x86_64-linux-opt --output rootfs/usr/local/bin/cvc4
+
+# Install Mathsat 5.6.3
+RUN curl -L https://mathsat.fbk.eu/download.php?file=mathsat-5.6.3-linux-x86_64.tar.gz | tar xz
+RUN cp mathsat-5.6.3-linux-x86_64/bin/mathsat rootfs/usr/local/bin
```

Comment:
I wasn't sure about licensing. If Mathsat is the only problematic one, that's not bad! I'll implement this.

Reply:
Strangely, I can't seem to easily find the MathSAT license so I can link to it. Can you point me in the right direction?

Reply:
They don't release the text of a specific license, as far as I know.  There's a copyright notice on their download page, which is essentially an "All Rights Reserved" notice.

https://mathsat.fbk.eu/download.html

Reply:
Yeah, the key sentence is "It can not be used in a commercial environment, particularly as part of a commercial product, without written permission.".

Reply:
We could ask Roberto...